routing/src/prefix.rs

```diff
@@ -350,6 +354,211 @@ impl<'de> Deserialize<'de> for Prefix {
+    type Output = PrefixSize;
+
+    fn add(self, other: PrefixSize) -> PrefixSize {
+        *self + other
+    }
+}
+
+impl Add<&PrefixSize> for &PrefixSize {
+    type Output = PrefixSize;
+
+    fn add(self, other: &PrefixSize) -> PrefixSize {
+        *self + *other
+    }
+}
+
+impl Sum<PrefixSize> for PrefixSize {
+    fn sum<I: Iterator<Item = PrefixSize>>(iter: I) -> Self {
+        iter.fold(PrefixSize::U128(0), |a, b| a + b)
+    }
+}
+
+impl<'a> Sum<&'a PrefixSize> for PrefixSize {
+    fn sum<I: Iterator<Item = &'a PrefixSize>>(iter: I) -> Self {
+        iter.fold(PrefixSize::U128(0), |a, b| a + b)
+    }
+}
+
+impl Sub<u128> for PrefixSize {
+    type Output = Self;
+
+    fn sub(self, int: u128) -> Self {
+        self - PrefixSize::U128(int)
+    }
+}
+
+impl Sub<u128> for &PrefixSize {
+    type Output = PrefixSize;
+
+    fn sub(self, int: u128) -> PrefixSize {
+        *self - int
+    }
+}
+
+impl Sub<PrefixSize> for PrefixSize {
+    type Output = Self;
+
+    fn sub(self, other: PrefixSize) -> Self {
+        match (self, other) {
+            (_, PrefixSize::U128(0)) => self,
+            (PrefixSize::U128(size_self), PrefixSize::U128(size_other)) => {
+                // May panic, just like a regular subtraction
+                PrefixSize::U128(size_self - size_other)
+            }
+            (PrefixSize::Ipv6MaxAddrs, PrefixSize::U128(size_other)) => {
+                PrefixSize::U128(u128::MAX - size_other + 1)
+            }
+            (PrefixSize::Ipv6MaxAddrs, PrefixSize::Ipv6MaxAddrs) => PrefixSize::U128(0),
+            (PrefixSize::U128(size_self), PrefixSize::Ipv6MaxAddrs) => {
+                // WILL panic, just like a regular subtraction
+                PrefixSize::U128(size_self - u128::MAX - 1)
+            }
+            _ => PrefixSize::Overflow,
+        }
+    }
+}
+
+impl Sub<&PrefixSize> for PrefixSize {
+    type Output = Self;
+
+    fn sub(self, other: &PrefixSize) -> Self {
+        self - *other
+    }
+}
+
+impl Sub<PrefixSize> for &PrefixSize {
+    type Output = PrefixSize;
+
+    fn sub(self, other: PrefixSize) -> PrefixSize {
+        *self - other
+    }
+}
+
+impl Sub<&PrefixSize> for &PrefixSize {
+    type Output = PrefixSize;
+
+    fn sub(self, other: &PrefixSize) -> PrefixSize {
+        *self - *other
+    }
+}
+
+impl PartialEq<u128> for PrefixSize {
+    fn eq(&self, other: &u128) -> bool {
+        match self {
+            PrefixSize::U128(size) => size == other,
+            _ => false,
+        }
+    }
+}
+
+impl PartialEq<PrefixSize> for u128 {
+    fn eq(&self, other: &PrefixSize) -> bool {
+        match other {
+            PrefixSize::U128(size) => size == self,
+            _ => false,
+        }
+    }
+}
+
+impl PartialOrd<u128> for PrefixSize {
+    fn partial_cmp(&self, other: &u128) -> Option<Ordering> {
+        match self {
+            PrefixSize::U128(size) => size.partial_cmp(other),
+            _ => Some(Ordering::Greater),
```

Comment:
I think the right answer here is to leave some elements incomparable, particular, `Overflow` is not greater than `Overflow` nor are they equal nor are they less than, thus the function should return none.  I think that, plus some careful treatment of the max value vs `u128::MAX` should allow the transitivity and duality property to hold.

Reply:
What happens if PrefixSize does not implement `Ord`?  Also, for an enum, what is the derived implementation of `Ord`, I'm curious?  I'm thinking that `Overflow` is a bit like `Nan`.